x-pack/legacy/plugins/monitoring/server/lib/calculate_rate.js

```diff
@@ -49,7 +49,16 @@ export function calculateRate(
 
   const earliestTimeInMillis = moment(timeWindowMin).valueOf();
   const latestTimeInMillis = moment(timeWindowMax).valueOf();
-  const millisDelta = latestTimeInMillis - earliestTimeInMillis;
+
+  // Effective latest and earliest timestamps depend not only on the time window bounds but also
+  // on whether we have data within those bounds. For example, if the time window bounds are "Last 1 hour"
+  // but we only have monitoring data for the last 25 minutes, we should use the latter for calculating
+  // the rate. If we only use the time window bounds, the rate will be calculated as lower than it
+  // should be.
+  const effectiveLatestTimeInMillis = Math.min(hitTimestampMoment, latestTimeInMillis);
+  const effectiveEarliestTimeInMillis = Math.max(earliestHitTimestampMoment, earliestTimeInMillis);
```

Comment:
Does it make sense to even care about `earliestTimeInMillis` at all? 

The query is bounded by `timeWindowMin` and `timeWindowMax` so
- `Math.min(hitTimestampMoment, latestTimeInMillis)` will always return `hitTimestampMoment`
- `Math.max(earliestHitTimestampMoment, earliestTimeInMillis)` will always return `earliestHitTimestampMoment`

Assuming that logic is true, we can just do:
`const millisDelta = hitTimestampMoment - earliestHitTimestampMoment;`

WDYT?

Reply:
Yeah, I actually had an implementation that simply removed the `timeWindowMin` and `timeWindowMax` arguments from the function signature. But then I wasn't sure if this might break anything so I chose this safer implementation.

If you're pretty confident that we'd never have a situation where this function is called such that `hitTimestampMoment` would be greater than `latestTimeInMillis` OR `earliestHitTimestampMoment` would be less than `earliestTimeInMillis`, then I agree with the simplification.

Reply:
Gotcha. I can't imagine a scenario, but what you have doesn't add any more confusion or complexity. Let's just leave it as is

Reply:
To be clear, I also would like to simplify this function if possible, safely.

I searched through the codebase and there are 6 existing calls to `calculateRate`. I'm checking if there's any possibility that any of them could be called with the scenario I described in my previous comment. If none of them turn out to be that way, I'll go ahead and simplify the implementation here.

Reply:
Yeah, in the case of all 6 calls the ES query that is being used to generate the docs (and therefore `earliestHitTimestampMoment` and `hitTimestampMoment`) is limiting the hits to within the bounds of the time picker (so `earliestTimeInMillis` and `latestTimeInMillis`). IOW, these queries should never produce docs such that `earliestHitTimestampMoment` would be less than `earliestTimeInMillis` OR `hitTimestampMoment` would be greater than `latestTimeInMillis`.

So I think it's safe to simplify as you suggested. I will make the change now.